python_apps/rtsp-camera-to-pravega-probe.py

```diff
@@ -0,0 +1,341 @@
+#
+
+import configargparse as argparse
+import ctypes
+import distutils.util
+import logging
+import os
+import signal
+import sys
+import time
+import traceback
+from threading import Thread, current_thread
+from http.server import HTTPServer, BaseHTTPRequestHandler
+
+import gi
+gi.require_version("Gst", "1.0")
+from gi.repository import GObject, Gst
+
+def bus_call(bus, message, loop):
+    """Callback for GStreamer bus messages"""
+    t = message.type
+    if t == Gst.MessageType.EOS:
+        logging.info("End-of-stream")
+        loop.quit()
+    elif t == Gst.MessageType.WARNING:
+        err, debug = message.parse_warning()
+        logging.warning("%s: %s" % (err, debug))
+    elif t == Gst.MessageType.ERROR:
+        err, debug = message.parse_error()
+        logging.error("%s: %s" % (err, debug))
+        loop.quit()
+    elif t == Gst.MessageType.ELEMENT:
+        details = message.get_structure().to_string()
+        logging.info("%s: %s" % (message.src.name, str(details),))
+    elif t == Gst.MessageType.PROPERTY_NOTIFY:
+        details = message.get_structure().to_string()
+        logging.debug("%s: %s" % (message.src.name, str(details),))
+    return True
+
+def on_queue_overrun(element):
+    logging.warning("Queue has overflowed and data has been lost. Try increasing buffer-size-mb.")
+
+def str2bool(v):
+    return bool(distutils.util.strtobool(v))
+
+def format_clock_time(ns):
+    """Format time in nanoseconds like 01:45:35.975000000"""
+    s, ns = divmod(ns, 1000000000)
+    m, s = divmod(s, 60)
+    h, m = divmod(m, 60)
+    return "%u:%02u:%02u.%09u" % (h, m, s, ns)
+
+def buffer_timestamp_probe(pad, info, data):
+    gst_buffer = info.get_buffer()
+    if gst_buffer:
+        data.update(gst_buffer.pts)
+        logging.info("buffer_timestamp_probe: %20s:%-8s: " % (
+            pad.get_parent_element().name,
+            pad.name) + data.to_string()
+        )
+        
+    else:
+        logging.info("buffer_timestamp_probe: %20s:%-8s: no buffer" % (
+            pad.get_parent_element().name,
+            pad.name))
+    return Gst.PadProbeReturn.OK
+
+class TimestampSet():
+    def __init__(self, tolerance):
+        self.value = 0
+        self.update_at = 0
+        self.update_tolerance = tolerance
+
+    def update(self, value):
+        self.value = value
+        self.update_at = int(time.time())
+    
+    def to_string(self):
+        return "value: %u at %u seconds since epoch" % (self.value, self.update_at)
+    
+    def is_uptodate(self):
+        return int(time.time()) - self.update_at < self.update_tolerance
+
+def ProbeHttpServer():
+    hostname = '0.0.0.0'
+    port = 8080
+    httpd = HTTPServer((hostname, port), ProbeHttpHandler)
+    def serve_forever(httpd):
+        with httpd:  # to make sure httpd.server_close is called
+            httpd.serve_forever()
+
+    thread = Thread(target=serve_forever, args=(httpd, ))
+    thread.setDaemon(True)
```

Comment:
Add a comment explaining why setDaemon is required.

Reply:
added